### PR TITLE
Implement PS-3879 (Encrypted temporary files)

### PR DESCRIPTION
**Pull request by percona-ysorokin**
https://jira.percona.com/browse/PS-3879

Introduced new global non-dynamic boolean variable 'encrypt_tmp_files' which,
when turned on, enables encryption of temporary files created for various
server needs (MyISAM 'OPTIMIZE' / 'REPAIR', filesort, etc.)
For each individual file a new random key is generated using 'my_rand_buffer()'
upon opening (no encryption / key storage plugin is involved in this process).
File data is encrypted using AES 256-bit algorithm in CBC block mode.

Cherry-picked commit MariaDB/server@d9340d6
"MDEV-8126 encryption for temp files"

Cherry-picked commit MariaDB/server@318c826
"always use my_b_pread() instead of mysql_file_pread()"

Cherry-picked commit MariaDB/server@6309a30
"my_b_fill, inline my_b_* functions instead of hairy macros"

Cherry-picked commit MariaDB/server@196e852
"misc IO_CACHE cleanups"

Cherry-picked commit MariaDB/server@1d21b22
"MDEV-10001 my_b_seek() may not work correctly after my_b_read() hits EOF"

Cherry-picked commit MariaDB/server@58b54b7
"MDEV-9044 : Binlog corruption in Galera"

Cherry-picked commit MariaDB/server@735a4a1
"MDEV-10508 Mariadb crash on out of disk space during dump import"

Cherry-picked commit MariaDB/server@1841557
"myisam/aria: don't mess with IO_CACHE::file"

Cherry-picked commit MariaDB/server@3a1c91d
"add P_S instrumentation to file operations in mf_iocache2.c"

Cherry-picked commit MariaDB/server@47d1679
"fix encryption.tempfiles to check that encrypt_tmp_files is ON"

Cherry-picked commit MariaDB/server@60dfe12
"MDEV-14868 MariaDB server crashes after using ROLLBACK TO when encrypt_tmp_files=ON"

Cherry-picked commit MariaDB/server@b75d845
"MDEV-14868 MariaDB server crashes after using ROLLBACK TO when encrypt_tmp_files=ON"

In contrast to MariaDB we preserve original 'my_b_fill()' implementation for
non-encrypted mode and use
'return _my_b_read(info, 0, 0) ? 0 : info->read_end - info->read_pos;'
only when 'MY_ENCRYPT' flag is set.

Applied experimental fix for MDEV-10259
"mysqld crash with certain statement length and order with Galera and encrypt-tmp-files=1"
(https://jira.mariadb.org/browse/MDEV-10259)
(http://lists.askmonty.org/pipermail/commits/2018-March/012144.html)
Fixed a problem in 'my_b_encr_read()' with updating 'read_end' and 'pos_in_file'.

Fixed problem in group_replication plugin code:
'Transaction_Message::append_cache()' method could call 'my_b_fill()' in case
when 'info->read_pos != info->read_end' which is not supported in the
encrypted mode.

Fixed "my_aes.h" interface to be plain C compatible.

Re-recorded 'main.mysqld--help-notwin' MTR test case because of the new
'--encrypt-tmp-files' system variable.

Fixed 'perfschema.show_sanity' MTR test case because of the new
'--encrypt-tmp-files' system variable.

Re-recorded 'perfschema.relaylog' and 'rpl.rpl_binlog_sender_fseek' MTR test
cases re-recorded because of the commit MariaDB/server@3a1c91d
"add P_S instrumentation to file operations in mf_iocache2.c".
Additional seek was already there, just not instrumented and
therefore, not reflected in statistics.

Added new 'sys_vars.enctypt_tmp_files_basic' MTR test case for the new
'encrypt_tmp_files' system variable.

Added 'mf_iocache-t' gunit test which tests basic 'io_cache' functionality
in encrypted / unencrypted mode.

Various formatting and spelling fixes.

### CODE REVIEW

**Comment by percona-ysorokin** [comment]
Jenkins build link
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1035

File: mysql-test/t/percona_encrypt_tmp_files.test [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# Row binlog format to fill binlog cache faster
+--source include/have_binlog_format_row.inc
+--source include/have_innodb.inc
+
+if (`SELECT @@global.encrypt_tmp_files = 0`)
+{
+   --skip Test requires --encrypt-tmp-files
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_encrypt_tmp_files.test]: --die ?

File: mysql-test/t/percona_encrypt_tmp_files.test [hunk 2]
```diff
@@ -0,0 +1,104 @@
+# Row binlog format to fill binlog cache faster
+--source include/have_binlog_format_row.inc
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_encrypt_tmp_files.test]: The test will be just skipped if the binlog format happens to be different (main suite does not use binlog format combinations) - maybe set the format in the testcase instead?

File: mysql-test/t/percona_encrypt_tmp_files.test [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+--source include/have_innodb.inc
+
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_encrypt_tmp_files.test]: --source include/have_myisam.inc

File: mysys/mf_iocache.c [hunk 4]
```diff
@@ -64,6 +65,18 @@ PSI_file_key key_file_io_cache;
...
 
+static int _my_b_cache_read(IO_CACHE *info, uchar *Buffer, size_t Count);
+static int _my_b_cache_read_r(IO_CACHE *info, uchar *Buffer, size_t Count);
+static int _my_b_seq_read(IO_CACHE *info, uchar *Buffer, size_t Count);
+static int _my_b_cache_write(IO_CACHE *info, const uchar *Buffer,
+                             size_t Count);
+static int _my_b_cache_write_r(IO_CACHE *info, const uchar *Buffer,
+                               size_t Count);
+
+static io_cache_encr_read_function _my_b_encr_read= 0;
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache.c]: NULL

File: mysys/mf_iocache.c [hunk 5]
```diff
@@ -64,6 +65,18 @@ PSI_file_key key_file_io_cache;
...
+static io_cache_encr_write_function _my_b_encr_write= 0;
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache.c]: NULL

File: mysys/mf_iocache.c [hunk 6]
```diff
@@ -97,6 +110,8 @@ static void
 init_functions(IO_CACHE* info)
 {
   enum cache_type type= info->type;
+  info->read_function= 0;                      /* Force a core if used */
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache.c]: NULL

File: mysys/mf_iocache.c [hunk 7]
```diff
@@ -97,6 +110,8 @@ static void
...
+  info->write_function= 0;                     /* Force a core if used */
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache.c]: NULL

File: mysys/mf_iocache.c [hunk 8]
```diff
@@ -106,21 +121,78 @@ init_functions(IO_CACHE* info)
...
+      break;
+    }
+    /* fall through */
+  case WRITE_CACHE:
+    if (info->myflags & MY_ENCRYPT)
+    {
+      info->write_function= _my_b_encr_write;
+      break;
+    }
+    /* fall through */
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache.c]: Why fall through here? Either read/write_function will be initialised already, either the DBUG_ASSERT below will fire

File: sql/binlog.cc [hunk 9]
```diff
@@ -1025,7 +1025,8 @@ class Binlog_event_writer
       {
         char checksum_buf[BINLOG_CHECKSUM_LEN];
         int4store(checksum_buf, checksum);
-        if (my_b_write(output_cache, checksum_buf, BINLOG_CHECKSUM_LEN))
+        if (my_b_write(output_cache, reinterpret_cast<uchar*>(checksum_buf),
```
Comment (laurynas-biveinis) [inline sql/binlog.cc]: static_cast not enough?

File: sql/filesort.cc [hunk 10]
```diff
@@ -1160,7 +1160,8 @@ write_keys(Sort_param *param, Filesort_info *fs_info, uint count,
       DBUG_RETURN(1);                           /* purecov: inspected */
   }
 
-  if (my_b_write(chunk_file, &merge_chunk, sizeof(merge_chunk)))
+  if (my_b_write(chunk_file, reinterpret_cast<uchar*>(&merge_chunk),
```
Comment (laurynas-biveinis) [inline sql/filesort.cc]: Likewise

File: sql/log_event.cc [hunk 11]
```diff
@@ -2170,7 +2170,7 @@ log_event_print_value(IO_CACHE *file, const uchar *ptr,
       struct timeval tm;
       my_timestamp_from_binary(&tm, ptr, meta);
       int buflen= my_timeval_to_str(&tm, buf, meta);
-      my_b_write(file, buf, buflen);
+      my_b_write(file, reinterpret_cast<uchar*>(buf), buflen);
```
Comment (laurynas-biveinis) [inline sql/log_event.cc]: Likewise

File: sql/mf_iocache_encr.cc [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+#include "my_aes.h"
+#include "my_rnd.h"
+#include "log.h"
+#include "mysql/psi/mysql_file.h"
+
+typedef struct {
+  ulonglong counter;
+  uint block_length, last_block_length;
+  uchar key[MY_AES_BLOCK_SIZE];
+  ulonglong inbuf_counter;
```
Comment (laurynas-biveinis) [inline sql/mf_iocache_encr.cc]: Move after "counter" to save some alignment padding bytes?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
This is only a superficial review.

Where are the encryption keys generated?

File: mysql-test/t/percona_encrypt_tmp_files.test (hunk 2 above)
Comment (percona-ysorokin) [reply mysql-test/t/percona_encrypt_tmp_files.test]: Good point

File: sql/binlog.cc (hunk 9 above)
Comment (percona-ysorokin) [reply sql/binlog.cc]: static_cast is not allowed for converting `char*` to `unsigned char*`

File: sql/filesort.cc (hunk 10 above)
Comment (percona-ysorokin) [reply sql/filesort.cc]: The same: `Merge_chunk*` and `uchar*` are considered to be `unrelated types`, so reinterpret_cast only

File: sql/mf_iocache_encr.cc [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+    if (mysql_file_read(info->file, wbuffer, wlength,
+                        info->myflags | MY_NABP))
+    {
+      info->error= -1;
+      DBUG_RETURN(1);
+    }
+
+    elength= wlength - (ebuffer - wbuffer);
+    set_iv(iv, pos_in_file, crypt_data->inbuf_counter);
+
```
Comment (robgolebiowski) [inline sql/mf_iocache_encr.cc]: IV should be encrypted first, before used. Based on this book https://books.google.pl/books?id=nnqdQsrZnJgC&lpg=PT141&ots=6GWXwmnDt8&dq=Bruce%20Schneier%20cbc%20or%20ctr&hl=pl&pg=PT133#v=onepage&q=Bruce%20Schneier%20cbc%20or%20ctr&f=false by Bruce Scheneier - point 4.3.4.3 The nonce should be encrypted before it should be used as an IV. Thus there need to be something like what Sergei included in his diff:
 my_aes_encrypt(iv, MY_AES_BLOCK_SIZE, eiv,
                 key, key_length, ecb_mode, NULL, false);
Where eiv - should be later used as IV.

**Review by robgolebiowski — CHANGES_REQUESTED** [review]

File: mysql-test/t/percona_encrypt_tmp_files.test (hunk 3 above)
Comment (percona-ysorokin) [reply mysql-test/t/percona_encrypt_tmp_files.test]: fixed

File: mysys/mf_iocache.c (hunk 4 above)
Comment (percona-ysorokin) [reply mysys/mf_iocache.c]: fixed

File: mysys/mf_iocache.c (hunk 5 above)
Comment (percona-ysorokin) [reply mysys/mf_iocache.c]: fixed

File: mysys/mf_iocache.c (hunk 6 above)
Comment (percona-ysorokin) [reply mysys/mf_iocache.c]: fixed

File: mysys/mf_iocache.c (hunk 7 above)
Comment (percona-ysorokin) [reply mysys/mf_iocache.c]: fixed

File: mysys/mf_iocache.c (hunk 8 above)
Comment (percona-ysorokin) [reply mysys/mf_iocache.c]: Reworked by separating "encrypted" / "non-encrypted" code paths to make it more readable.

File: sql/log_event.cc (hunk 11 above)
Comment (percona-ysorokin) [reply sql/log_event.cc]: No cast is needed, `buf` is already `uchar*` here.

File: mysql-test/t/percona_encrypt_tmp_files.test (hunk 2 above)
Comment (percona-ysorokin) [reply mysql-test/t/percona_encrypt_tmp_files.test]: Reworked

File: mysql-test/t/percona_encrypt_tmp_files.test (hunk 1 above)
Comment (percona-ysorokin) [reply mysql-test/t/percona_encrypt_tmp_files.test]: Changed to `--die`

File: sql/mf_iocache_encr.cc (hunk 12 above)
Comment (percona-ysorokin) [reply sql/mf_iocache_encr.cc]: Internal algorithm relies on the fact that `inbuf_counter` is the last member of the struct. Moreover, in this particular case there is no padding gaps:
```c++
ulonglong counter; // 8 bytes
uint block_length, last_block_length; // 4 + 4 = 8 bytes
uchar key[MY_AES_BLOCK_SIZE]; // 16 bytes
ulonglong inbuf_counter; // 8 bytes
```

**Comment by percona-ysorokin** [comment]
Jenkins build links:
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1742/ (normal)
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1743/ (MTR args: --mysqld=--encrypt-tmp-files)

**Comment by percona-ysorokin** [comment]
@laurynas-biveinis @robgolebiowski a new commit has been pushed to this PR and is ready for review.

File: include/my_sys.h [hunk 14]
```diff
@@ -487,46 +488,85 @@ typedef void (*my_error_reporter)(enum loglevel level, const char *format, ...)
 
 extern my_error_reporter my_charset_error_reporter;
 
-/* defines for mf_iocache */
+/* inline functions for mf_iocache */
```
Comment (laurynas-biveinis) [inline include/my_sys.h]: Consider using MY_NODISCARD for all the changed functions (if it does not cause warnings in legacy code)

File: mysql-test/suite/binlog/t/binlog_check_flush_log_assertion.test [hunk 15]
```diff
@@ -50,7 +50,6 @@ SET GLOBAL debug='+d,error_unique_log_filename';
 RESET MASTER;
 
 SET GLOBAL debug='-d,error_unique_log_filename';
-SET @@GLOBAL.DEBUG= @debug_save;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/binlog/t/binlog_check_flush_log_assertion.test]: Why?

File: mysql-test/t/percona_encrypt_tmp_files.test [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+CREATE TABLE t2 (a TEXT) ENGINE=InnoDB;
+START TRANSACTION;
+INSERT t2 SELECT REPEAT(seq, 1000) FROM seq_1_to_15;
+SAVEPOINT foo;
+INSERT t2 SELECT REPEAT(seq, 1000) FROM seq_16_to_30;
+ROLLBACK TO SAVEPOINT foo;
+INSERT t2 SELECT REPEAT(seq, 1000) FROM seq_31_to_40;
+COMMIT;
+
+# SHOW BINLOG EVENTS must sucseed without 'CORRUPTED BINLOG' error
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_encrypt_tmp_files.test]: succeed

File: mysql-test/t/percona_encrypt_tmp_files.test [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+connection default;
+
+FLUSH BINARY LOGS;
+
+DROP TABLE t1, t2;
+
+SET GLOBAL binlog_cache_size = @saved_binlog_cache_size;
+
+let $MYSQLD_DATADIR = `SELECT @@datadir`;
+exec $MYSQL_BINLOG $MYSQLD_DATADIR/master-bin.000001 | $MYSQL;
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_encrypt_tmp_files.test:110]: I see that there is a RESET MASTER above, but make sure to test with --repeat=3 at least

File: mysys/mf_iocache.c [hunk 18]
```diff
@@ -72,7 +85,7 @@ PSI_file_key key_file_io_cache;
     info		IO_CACHE handler
 
   NOTES
-    This is called on automaticly on init or reinit of IO_CACHE
+    This is called on automatically on init or reinit of IO_CACHE
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache.c:94]: Drop :(

File: mysys/mf_iocache.c [hunk 19]
```diff
@@ -1723,7 +1850,7 @@ int main(int argc, char** argv)
   /* check correctness of tests */
   if (total_bytes != status.st_size)
   {
-    fprintf(stderr,"Not the same number of bytes acutally  in file as bytes \
+    fprintf(stderr,"Not the same number of bytes actually  in file as bytes \
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache.c]: might as well remove the extra space between "actually" and "in"

File: mysys/mf_iocache2.c [hunk 20]
```diff
@@ -182,7 +183,6 @@ void my_b_seek(IO_CACHE *info,my_off_t pos)
   DBUG_VOID_RETURN;
 }
 
-
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache2.c]: Drop

File: sql/binlog.cc [hunk 21]
```diff
@@ -5922,7 +5922,10 @@ bool MYSQL_BIN_LOG::reset_logs(THD* thd, bool delete_only)
     }
     // don't clear global_sid_map because it's used by the relay log too
     if (gtid_state->init() != 0)
+    {
```
Comment (laurynas-biveinis) [inline sql/binlog.cc]: Why? Is this something that should be reported upstream as a code quality bug at least?

File: sql/binlog.cc [hunk 22]
```diff
@@ -5936,11 +5939,12 @@ bool MYSQL_BIN_LOG::reset_logs(THD* thd, bool delete_only)
...
       goto err;
   }
-  my_free((void *) save_name);
 
 err:
-  if (error == 1)
+  if (error)
     name= const_cast<char*>(save_name);
+  else
+    my_free(const_cast<char *>(save_name));
```
Comment (laurynas-biveinis) [inline sql/binlog.cc]: Why? Is this something that should be reported upstream as a code quality bug at least?

File: sql/log_event.cc (hunk 11 above)
Comment (laurynas-biveinis) [reply sql/log_event.cc]: buf is not uchar* ?

File: sql/mf_iocache_encr.cc [hunk 23]
```diff
@@ -0,0 +1,296 @@
+/*
+   Copyright (c) 2018, Percona Inc. All Rights Reserved.
```
Comment (laurynas-biveinis) [inline sql/mf_iocache_encr.cc]: Is this correct copyright notice for a MariaDB port? Or is this file not a MariaDB port?

File: mysys/mf_iocache.c [hunk 24]
```diff
@@ -64,6 +65,18 @@ PSI_file_key key_file_io_cache;
 #define IO_ROUND_UP(X) (((X)+IO_SIZE-1) & ~(IO_SIZE-1))
 #define IO_ROUND_DN(X) ( (X)            & ~(IO_SIZE-1))
 
+static int _my_b_cache_read(IO_CACHE *info, uchar *Buffer, size_t Count);
```
Comment (laurynas-biveinis) [inline mysys/mf_iocache.c]: MY_NODISCARD all of them

File: sql/mf_iocache_encr.cc [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+  compile_time_assert(sizeof(iv) >= sizeof(position) + sizeof(counter));
+
+  io_cache_aes_block_buffer nonce_block;
+  compile_time_assert(sizeof(nonce_block) == MY_AES_BLOCK_SIZE);
+
+  memset(nonce_block, 0, sizeof(nonce_block));
+  memcpy(nonce_block, &position, sizeof(position));
+  memcpy(&nonce_block[sizeof(position)], &counter, sizeof(counter));
+
+  int encryption_result MY_ATTRIBUTE((unused))= my_aes_encrypt(nonce_block,
```
Comment (laurynas-biveinis) [inline sql/mf_iocache_encr.cc:79]: Consider checking in release build too

File: sql/sql_parse.cc [hunk 26]
```diff
@@ -4456,7 +4456,9 @@ case SQLCOM_PREPARE:
       if (!res)
         my_ok(thd);
     } 
-    
+    else
+      res= 1; // reload_acl_and_cache failed
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: Why?

File: mysql-test/suite/rpl/t/rpl_binlog_sender_fseek.test [hunk 27]
```diff
@@ -9,9 +9,9 @@
 # sometimes, cross the IO_CACHE block size. After running the workload, the test
 # case will verify the P_S counter COUNT_MISC for "wait/io/file/sql/binlog"
 # EVENT_NAME at performance_schema.file_summary_by_event_name. In a single
-# binary log file case, the counter should not be more than 5 (as there are some
+# binary log file case, the counter should not be more than 6 (as there are some
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl/t/rpl_binlog_sender_fseek.test]: Report the bug of PFS not registering seeks upstream

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
This is going to be maintenance nightmare for IO_CACHE changes, but I don't see a better alternative

File: mysql-test/suite/binlog/t/binlog_check_flush_log_assertion.test (hunk 15 above)
Comment (percona-ysorokin) [reply mysql-test/suite/binlog/t/binlog_check_flush_log_assertion.test]: Rolled back

File: mysql-test/suite/rpl/t/rpl_binlog_sender_fseek.test (hunk 27 above)
Comment (percona-ysorokin) [reply mysql-test/suite/rpl/t/rpl_binlog_sender_fseek.test]: Reported PS-3937 "Some file operations in mf_iocache2.c are not instrumented"
https://jira.percona.com/browse/PS-3937

File: mysql-test/t/percona_encrypt_tmp_files.test (hunk 16 above)
Comment (percona-ysorokin) [reply mysql-test/t/percona_encrypt_tmp_files.test]: fixed

File: mysql-test/t/percona_encrypt_tmp_files.test (hunk 17 above)
Comment (percona-ysorokin) [reply mysql-test/t/percona_encrypt_tmp_files.test:110]:
```
./mysql-test/mtr --debug-server percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files --parallel=8 --repeat=16
```
runs fine

File: mysys/mf_iocache.c (hunk 18 above)
Comment (percona-ysorokin) [reply mysys/mf_iocache.c:94]: ???

Comment (laurynas-biveinis) [reply mysys/mf_iocache.c:94]: It's a comment typo fix. Only adds to diff maintenance

File: mysys/mf_iocache.c (hunk 19 above)
Comment (percona-ysorokin) [reply mysys/mf_iocache.c]: fixed

File: mysys/mf_iocache2.c (hunk 20 above)
Comment (percona-ysorokin) [reply mysys/mf_iocache2.c]: done

File: sql/binlog.cc (hunk 21 above)
Comment (percona-ysorokin) [reply sql/binlog.cc]: Removed from this PR, will be reported later.

File: sql/binlog.cc (hunk 22 above)
Comment (percona-ysorokin) [reply sql/binlog.cc]: Removed from this PR, will be reported later.

File: sql/log_event.cc (hunk 11 above)
Comment (percona-ysorokin) [reply sql/log_event.cc]: Fixed

File: sql/mf_iocache_encr.cc (hunk 23 above)
Comment (percona-ysorokin) [reply sql/mf_iocache_encr.cc]: Fixed

File: sql/sql_parse.cc (hunk 26 above)
Comment (percona-ysorokin) [reply sql/sql_parse.cc]: Reverted

File: sql/mf_iocache_encr.cc (hunk 25 above)
Comment (percona-ysorokin) [reply sql/mf_iocache_encr.cc:79]: There is almost no change AES encrypt function can fail in ECB mode for a single block.

**Comment by percona-ysorokin** [comment]
Rebased on the top of the latest 5.7 which includes IO_CACHE refactoring.

Jenkins build link (normal):
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1819/

Jenkins build link (--mysqld=--encrypt-tmp-files):
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1826/

`sys_vars.encrypt_tmp_files_basic` in the second run fails constantly because of the default sys var value
changed with `--mysqld=--encrypt-tmp-files`.

**Comment by percona-ysorokin** [comment]
@laurynas-biveinis The PR is ready for another round of review.

File: mysql-test/t/percona_encrypt_tmp_files.test [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+INSERT INTO  seq_31_to_40 SELECT seq FROM universal_seq WHERE seq <= 40 AND seq >= 31 ORDER BY seq;
+
+#
+# filesort, my_b_pread, seeks in READ_CACHE
+#
+CREATE TABLE t1(v VARCHAR(10), c CHAR(10), t TEXT, KEY(v), KEY(c), KEY(t(10)));
+INSERT INTO t1(v) SELECT CONCAT(CHAR(ASCII('a') + s2.seq), REPEAT(' ', s1.seq))
+  FROM seq_0_to_9 AS s1, seq_0_to_26 AS s2;
+UPDATE t1 SET c = v, t = v;
+SELECT sql_big_result t, COUNT(t) FROM t1 GROUP BY t LIMIT 10;
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_encrypt_tmp_files.test:55]: sql_big_result uppercase

File: mysys/mf_iocache.c (hunk 18 above)
Comment (laurynas-biveinis) [reply mysys/mf_iocache.c:94]: I suppose you want to keep it then :)

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**